### PR TITLE
Code Modernization: Replace strpos() with str_starts_with()

### LINKED ISSUE #32854: Code modernization: Replace usage of `strpos` with `str_starts_with`
Source: https://core.trac.wordpress.org/ticket/58012

Let's follow Core's lead once Jetpack requires WordPress 6.3.

Primary issue: #32865

### PR DESCRIPTION
**Pull request by jeherve**
Fixes #32854

## Proposed changes:

`str_starts_with()` was introduced in PHP 8.0 to perform a case-sensitive check indicating if the string to search in (haystack) begins with the given substring (needle).

WordPress core includes a polyfill for `str_starts_with()` on PHP < 8.0 as of WordPress 5.9.

This commit replaces usage of `strpos()` with `str_starts_with()`, making the code more readable and consistent, as well as improving performance in PHP 8+ environments.

While `strpos()` is slightly faster than the polyfill on PHP < 8.0, `str_starts_with()` is noticeably faster on PHP 8.0+, as it is optimized to avoid unnecessarily searching along the whole haystack if it does not find the needle.

> **Note**
> 
> This doesn't update uses of `strpos()` in a few places:
> - The autoloader since it can be used outside of WordPress.
> - The VaultPress plugin since it can still be used on older versions of WordPress, where the `str_starts_with()` shim did not exist.

Matching core change: https://core.trac.wordpress.org/ticket/58012

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* Primary issue: #32865

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

* Activate all touched plugins and watch out for fatals on the frontend and wp-admin.

### CODE REVIEW

**Comment by kraftbj** [comment]
@jeherve A few of the tests need mocks for `str_starts_with` or something similar.

**Comment by jeherve** [comment]
Good point. I just pushed 81955927e5d2c021fabcfb2702b575aae440e9b1 to try to fix that.

File: projects/packages/assets/src/class-assets.php [hunk 1]
```diff
@@ -450,9 +450,9 @@ public static function wp_default_scripts_hook( $wp_scripts ) {
 		$content_dir = Jetpack_Constants::get_constant( 'WP_CONTENT_DIR' );
 		$abspath     = Jetpack_Constants::get_constant( 'ABSPATH' );
 
-		if ( strpos( $lang_dir, $content_dir ) === 0 ) {
+		if ( \str_starts_with( $lang_dir, $content_dir ) ) {
```
Comment (anomiex) [inline projects/packages/assets/src/class-assets.php]: I'm not sure that `wp-includes/compat.php` winds up getting loaded from the `wp-admin/load-scripts.php` entry point where this hook gets called from.

File: projects/packages/assets/tests/php/patchwork.json [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-	"redefinable-internals": [ "filemtime", "is_readable" ]
+	"redefinable-internals": [ "filemtime", "is_readable", "str_starts_with" ]
```
Comment (anomiex) [inline projects/packages/assets/tests/php/patchwork.json]: No need to mock it, just polyfill it in bootstrap.php or something like

```
if ( ! function_exists( 'str_starts_with' ) ) {
    function str_starts_with( ...
}
```

File: projects/packages/assets/src/class-assets.php (hunk 1 above)
Comment (anomiex) [reply projects/packages/assets/src/class-assets.php]: I note that `wp-admin/load-styles.php` has [a comment calling out that lack](https://github.com/WordPress/wordpress-develop/blob/0b8ca16ea3bd9722bd1a38f8ab68901506b1a0e7/src/wp-admin/load-styles.php#L76).

File: projects/packages/redirect/tests/php/patchwork.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+	"redefinable-internals": [ "str_starts_with" ]
```
Comment (anomiex) [inline projects/packages/redirect/tests/php/patchwork.json]: Same, no need to mock, just polyfill in bootstrap.php or something

File: projects/packages/assets/tests/php/patchwork.json (hunk 2 above)
Comment (jeherve) [reply projects/packages/assets/tests/php/patchwork.json]: 👍 4eaadb626f7a0b6e834a5534b80138beac8d4ae1

File: projects/packages/redirect/tests/php/patchwork.json (hunk 3 above)
Comment (jeherve) [reply projects/packages/redirect/tests/php/patchwork.json]: 👍 4eaadb626f7a0b6e834a5534b80138beac8d4ae1

File: projects/packages/assets/src/class-assets.php (hunk 1 above)
Comment (jeherve) [reply projects/packages/assets/src/class-assets.php]: That makes sense. c020b9f9e8ea93f414af7b966a86cb7ddd1715dc

File: projects/packages/assets/changelog/update-strpos-str-starts-with [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: changed
+
+Code Modernization: Replace usage of strpos() with str_starts_with().
```
Comment (anomiex) [inline projects/packages/assets/changelog/update-strpos-str-starts-with]:
```suggestion
Comment: Comment why we don't use str_starts_with() in Assets::wp_default_scripts_hook. No change to functionality.

```

**Review by anomiex — COMMENTED** [review]
Needs a trunk merge, otherwise it looks ok to me

**Review by anomiex — APPROVED** [review]

**Pull request merged**